{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reformat the mutation data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Gives Mutation name and locations\n",
    "* Mutation occurences in each sample \n",
    "* Consider mutation at each locus a mutation (doesn't matter the type)\n",
    "* TCGA barcode: 'TCGA-W5-AA2I-11A-11R-A41I-07',\n",
    " 'TCGA-W5-AA2I-01A-32R-A41I-07' , 11A is normal, 01A is tumor\n",
    "* Assume that the normal tissue has no mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#Filter the mutation samples and epxpression sample:\n",
    "#if sample in expression:\n",
    "#        if sample is normal (larger than ) Tumor types range from 01 - 09, normal types from 10 - 19:\n",
    "#            add this sample to final set of samples\n",
    "#        else:\n",
    "#            dont add this smaple to final set\n",
    "#if sample in mutation data:\n",
    "#    if sample in mutation and not in expression:\n",
    "#        dont add this sample\n",
    "#    else:\n",
    "#        add this sample to final sample set\n",
    "#\n",
    "\n",
    "def get_final_sample_set(geneexpression_file,mutation_file):\n",
    "#Filter the mutation samples and epxpression sample:\n",
    "#if sample in expression:\n",
    "#        if sample is normal (larger than 9) #Tumor types range from 01 - 09, normal types from 10 - 19:\n",
    "#            add this sample to final set of samples\n",
    "#        else:\n",
    "#            dont add this smaple to final set\n",
    "#if sample in mutation data:\n",
    "#    if sample in mutation and not in expression:\n",
    "#        dont add this sample\n",
    "#    else:\n",
    "#        add this sample to final sample set\n",
    "    \n",
    "    samples_with_expression = open(geneexpression_file).readline().strip().split('\\t')[1:]\n",
    "    for i in range(len(samples_in_expression)):\n",
    "        samples_with_expression[i] = '-'.join(samples_with_expression[i].split('-')[:4])\n",
    "        \n",
    "    samples_with_mutation = {}\n",
    "    for line in open(mutation_file):\n",
    "        s = '-'.join(line.split('\\t')[0].split('-')[:4])\n",
    "        samples_with_mutation[s] = 0\n",
    "    #print samples_with_mutation\n",
    "    \n",
    "    finalsampleset = {}\n",
    "    for s1 in samples_with_expression:\n",
    "        code = 0\n",
    "        try:\n",
    "            code = int (s1.split('-')[-1][:2])\n",
    "        except:\n",
    "            print \"ERROR\",s1\n",
    "        if  code >9:\n",
    "            finalsampleset[s1] = 1\n",
    "        else:\n",
    "            continue\n",
    "    for s2 in samples_with_mutation:\n",
    "        if s2 in samples_with_expression:\n",
    "            finalsampleset[s2] = 1\n",
    "   \n",
    "    return finalsampleset.keys()\n",
    "\n",
    "def istumor(sample):\n",
    "    if int (sample.split('-')[-1][:2]) > 9:\n",
    "        return False\n",
    "    return True\n",
    "\n",
    "def makecovariate(finalsampleset,outfile):\n",
    "    out = open(outfile,'w')\n",
    "    line = \"id\"\n",
    "    for s in sorted(finalsampleset):\n",
    "        line +='\\t'+s\n",
    "    out.write(line+'\\n')\n",
    "    line= \"mutation\"\n",
    "    for s in sorted(finalsampleset):\n",
    "        if istumor(s):\n",
    "            line +='\\t'+'1'\n",
    "        else:\n",
    "            line +='\\t'+'0'\n",
    "    out.write(line+'\\n')\n",
    "    out.close()\n",
    "    \n",
    "    return\n",
    "\n",
    "def format_mutation_and_genotype(mutationinputfile,tag,snp_location_out,genotype_out,finalsampleset):   \n",
    "    file = open(mutationinputfile).read().split('\\n')\n",
    "    print len(file)\n",
    "    mutations = {}\n",
    "    samples = {}\n",
    "    for line in file:\n",
    "        tmp = line.strip().split('\\t')\n",
    "        if len(tmp)==1:\n",
    "            continue\n",
    "        mut = tmp[1]+'_'+tmp[2]\n",
    "        sam = '-'.join(tmp[0].split('-')[:4])\n",
    "        if mut not in mutations:\n",
    "            mutations[mut] = 1\n",
    "        else:\n",
    "            mutations[mut] += 1\n",
    "        if sam not in samples:\n",
    "            samples[sam] = {mut:1}\n",
    "        else:\n",
    "            samples[sam][mut] = 1\n",
    "    print len(samples)\n",
    "    \n",
    "    #mutations[\"chrX_1\"]=1\n",
    "    translatedict = {}\n",
    "    for i in range(1,24):\n",
    "        c = 'chr'+str(i)\n",
    "        translatedict[c] =c \n",
    "    #translatedict\n",
    "    \n",
    "    \n",
    "    #create a mutationname dict as well:\n",
    "    mutnames = {}\n",
    "\n",
    "    #make file SNP_locations:\n",
    "    target = open(snp_location_out,'w')\n",
    "    target.write(\"snp\\tchr\\tpos\")\n",
    "    mutationlocs = []\n",
    "    factor = 10**12\n",
    "    for mut in mutations:\n",
    "        tmp = mut.split('_')\n",
    "        try:\n",
    "            loc = int(tmp[0].replace('chr',''))*factor + int(tmp[1])\n",
    "        except:\n",
    "            if 'x' in tmp[0] or 'X' in tmp[0]:\n",
    "                loc = 25*factor + int(tmp[1]) #chrx is chr25\n",
    "            elif 'y' in tmp[0] or 'Y' in tmp[0]:\n",
    "                loc = 26*factor + int(tmp[1]) #chrx is chr26\n",
    "            elif 'm' in tmp[0] or 'M' in tmp[0]:\n",
    "                loc = 27*factor + int(tmp[1]) #chrx is chr27\n",
    "        mutationlocs += [loc]\n",
    "    mutationlocs = sorted(mutationlocs)\n",
    "    for i in range(len(mutationlocs)):\n",
    "        loc = mutationlocs[i]\n",
    "        chromname = \"chr\"+str(loc/factor)\n",
    "        try:\n",
    "            chromname = translatedict[chromname]\n",
    "        except:\n",
    "            if chromname == 'chr25':\n",
    "                chromname = 'chrX'\n",
    "            elif chromname == 'chr26':\n",
    "                chromname = 'chrY'\n",
    "            elif chromname == 'chr27':\n",
    "                chromname = 'chrM'\n",
    "            else:\n",
    "                print \"ERROR\"\n",
    "                continue\n",
    "        mutname = \"snp_\"+tag+\"_\"+str(i+1)\n",
    "        mutnames[mutname] = chromname+'_'+str(loc%factor)\n",
    "        line = mutname +'\\t'+chromname+'\\t'+str(loc%factor)\n",
    "\n",
    "        #print line\n",
    "        target.write('\\n'+line)\n",
    "    target.close()\n",
    "    \n",
    "    ### Printing the genotype file\n",
    "    #samplenames = sorted(samples.keys())\n",
    "    samplenames = sorted(finalsampleset)\n",
    "    target = open(genotype_out,'w')\n",
    "    line = \"id\"\n",
    "    for sam in samplenames:\n",
    "        line += '\\t'+'-'.join(sam.split(\"-\")[:4])\n",
    "    target.write(line)\n",
    "    for i in range(len(mutnames)):\n",
    "        mutname = \"snp_\"+tag+\"_\"+str(i+1)\n",
    "        vector = [0]*len(samplenames)\n",
    "        mutlocname = mutnames[mutname]\n",
    "        for i in range(len(samplenames)):\n",
    "            sam = samplenames[i]\n",
    "            if sam in samples:\n",
    "                if mutlocname in samples[sam]:\n",
    "                    vector[i] = 1\n",
    "        line = mutname+'\\t'+'\\t'.join(map(str,vector))\n",
    "        target.write('\\n'+line)\n",
    "        #print mutname\n",
    "        #print vector\n",
    "    target.close()\n",
    "    \n",
    "    \n",
    "def format_geneloc_and_geneexpression(ucsc_loc_file,expressionfile,tag,genelocation_out,geneexpression_out,finalsampleset):\n",
    "    \n",
    "    # get the locations of the gene existing in the file. \n",
    "    locationfile = ucsc_loc_file\n",
    "    genelocdict = {}\n",
    "    #count = 0\n",
    "    for line in open(locationfile):\n",
    "        #count +=1 \n",
    "        #if count == 10:\n",
    "        #    break\n",
    "        #print line\n",
    "        tmp = line.split('\\t')\n",
    "        #print tmp[8]\n",
    "        genelocdict[tmp[7]] = [tmp[1],tmp[3]]\n",
    "        \n",
    "        genes = {}\n",
    "    output1 = open(genelocation_out,'w')\n",
    "    output2 = open(\"geneexpression_out.tmp\",'w')\n",
    "\n",
    "    line = \"id\\t\"+'\\t'.join(open(expressionfile).readline().split('\\t')[1:])\n",
    "    output2.write(line)\n",
    "\n",
    "\n",
    "    output1.write(\"geneid\\tchr\\ts1\\ts2\\n\")\n",
    "    #samples_in_expression:\n",
    "    gene_nointlocationcount = 0\n",
    "    for line in open(expressionfile):\n",
    "        tmp = line.strip().split('\\t')\n",
    "        gene = tmp[0].split('|')[0]\n",
    "        if gene in genelocdict:\n",
    "            line1 = gene+'\\t'+'\\t'.join(genelocdict[gene])+'\\t'+genelocdict[gene][1]\n",
    "            output1.write(line1+'\\n')\n",
    "            line2 = gene+'\\t'+'\\t'.join(tmp[1:])\n",
    "            output2.write(line2+'\\n')\n",
    "        else: \n",
    "            gene_nointlocationcount += 1\n",
    "    print \"genes not found in ucsc list\",gene_nointlocationcount\n",
    "    output1.close()\n",
    "    output2.close()\n",
    "    \n",
    "    #filter the gene expression another time. \n",
    "    # get only samples from the final set\n",
    "    data = pd.read_csv(\"geneexpression_out.tmp\",sep = '\\t',index_col=0)\n",
    "    columnames = list(data)\n",
    "    rownames = list(data.index)\n",
    "\n",
    "    finaldata = {}\n",
    "    for col in columnames:\n",
    "\n",
    "        s = '-'.join(col.strip().split('-')[:4])\n",
    "        #print col,s\n",
    "        if s in finalsampleset:\n",
    "            finaldata[s] = list(data[col])\n",
    "        else:\n",
    "            continue\n",
    "    output3 = open(geneexpression_out,'w')\n",
    "    line = 'id'\n",
    "    for s in sorted(finalsampleset):\n",
    "        line += '\\t' + s\n",
    "    #print line\n",
    "    output3.write(line+'\\n')\n",
    "    numline = len(finaldata[finalsampleset[0]])\n",
    "    for i in range(numline):\n",
    "        line = rownames[i]\n",
    "        #print line \n",
    "        for s in sorted(finalsampleset):\n",
    "            #print finaldata[s][i]\n",
    "            line += '\\t' + str(finaldata[s][i])\n",
    "        output3.write(line+'\\n')\n",
    "    output3.close()\n",
    "    \n",
    "    return\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#INPUTS\n",
    "geneexpression_file = \"/Users/vungo/Work/gdac.broadinstitute.org_CHOL.Merge_rnaseqv2__illuminahiseq_rnaseqv2__unc_edu__Level_3__RSEM_genes_normalized__data.Level_3.2016012800.0.0/CHOL.rnaseqv2__illuminahiseq_rnaseqv2__unc_edu__Level_3__RSEM_genes_normalized__data.data.txt\"\n",
    "mutationfile = \"/Users/vungo/Work/gdac.broadinstitute.org_CHOL-TP.Mutation_CHASM.Level_4.2016012800.0.0/mutationslist.txt\"\n",
    "ucsc_loc_file = \"/Users/vungo/Google_Drive/UCSC_gene_loc_and_names.txt\"\n",
    "tag = \"testCHOL\"\n",
    "\n",
    "finalsampleset = get_final_sample_set(geneexpression_file,mutationfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5282\n",
      "35\n",
      "genes not found in ucsc list 1668\n"
     ]
    }
   ],
   "source": [
    "\n",
    "snp_locationoutut = \"/Users/vungo/Work/testSNP_locations.txt\"\n",
    "genotypeout = \"/Users/vungo/Work/testGenotype.txt\"\n",
    "format_mutation_and_genotype(mutationfile,tag,snp_locationoutut,genotypeout,finalsampleset)\n",
    "\n",
    "genelocation_out = \"/Users/vungo/Work/testGenelocations.txt\"\n",
    "geneexpression_out = \"/Users/vungo/Work/testGeneexpreesion.txt\"\n",
    "\n",
    "format_geneloc_and_geneexpression(ucsc_loc_file,geneexpression_file,tag,genelocation_out,geneexpression_out,finalsampleset)\n",
    "\n",
    "#make covariate\n",
    "covariatefile = \"/Users/vungo/Work/testCovariate.txt\"\n",
    "makecovariate(finalsampleset,covariatefile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
